pyasice/xmlsig.py

```diff
@@ -0,0 +1,509 @@
+        :param hash_type: the hash function to use for digesting
+        :return:
+        """
+        try:
+            digest_algo = self.DIGEST_ALGORITHMS[hash_type]
+        except KeyError as e:
+            raise ValueError("Unknown hash type: %s" % hash_type) from e
+
+        digest_fn = getattr(hashlib, hash_type)
+        doc_hash = digest_fn(binary_data).digest()
+
+        signed_info = self._get_node("ds:SignedInfo")
+        first_ref_entry = signed_info.find('.//ds:Reference[@Type=""]', self.NAMESPACES)
+        doc_id = first_ref_entry.attrib["Id"]
+
+        doc_props = self._get_node("xades:SignedDataObjectProperties")
+        first_doc_entry = doc_props.find('.//xades:DataObjectFormat[@ObjectReference="#%s"]' % doc_id, self.NAMESPACES)
+
+        if self.doc_ids:
+            next_num = len(self.doc_ids) + 1
+
+            # generate new Id attribute
+            while True:
+                new_doc_id = "r-id-{}".format(next_num)
+                if new_doc_id not in self.doc_ids:
+                    break
+                next_num += 1
+
+            # Instead of manually creating elements, just copy the structure
+            new_ref_entry = copy.deepcopy(first_ref_entry)
+            signed_info.append(new_ref_entry)
+
+            new_doc_entry = copy.deepcopy(first_doc_entry)
+            doc_props.append(new_doc_entry)
+        else:
+            new_doc_id = doc_id.format(DOCUMENT_NUMBER=1)
+            new_doc_entry = first_doc_entry
+            new_ref_entry = first_ref_entry
+
+        self.doc_ids.append(new_doc_id)
+        new_ref_entry.attrib["Id"] = new_doc_id
+        new_ref_entry.attrib["URI"] = file_name
+        new_ref_entry.find(".//ds:DigestMethod", self.NAMESPACES).attrib["Algorithm"] = digest_algo
+        new_ref_entry.find(".//ds:DigestValue", self.NAMESPACES).text = base64.b64encode(doc_hash)
+
+        new_doc_entry.attrib["ObjectReference"] = "#%s" % new_doc_id
+        new_doc_entry.find(".//xades:MimeType", self.NAMESPACES).text = mime_type
+
+        return self
+
+    def update_signed_info(self):
+        """Calculate the digest over SignedProperties and embed it in SignedInfo"""
+
+        # Set signing time
+        time_node = self._get_signing_time_node()
+        # Add a UTC timestamp. Can't use isoformat() as it adds +00:00 and microseconds
+        #  which can break the parser elsewhere
+        time_node.text = get_utc_time().strftime("%Y-%m-%dT%H:%M:%SZ")
+
+        self._calc_signed_properties_hash(update=True)
+
+        self._prepared = True
+        return self
+
+    @property
+    def prepared(self):
+        if self._prepared is None:
+            self._prepared = False
+            if self.get_signing_time():
+                old, new = self._calc_signed_properties_hash()
+                if old == new:
+                    self._prepared = True
+                else:
+                    raise ValueError(f"{old}\n{new}")
```

Comment:
Tried running the first block of test. Naturally it failed as there wasn't any real signing done. But on `container.verify_signatures()` it raised this error. `ValueError: None zUt0mQBBGpxQk3WF13hQrUrRuYk7rT+2J6LhsBxKUk4=`
Not very descriptive.

Reply:
It was a leftover from debugging. Fixed. Thanks :)